### PR TITLE
Institute parser for Core.

### PR DESCRIPTION
**Pull request by patrickt**
I tried to pull the history from this patch over but I was not able to
figure out how to resolve conflicts from `git am`.

### CODE REVIEW

File: semantic-core/src/Data/Core/Parser.hs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  plain = Token.ident coreIdents
+  quote = between (string "#{") (symbol "}") (fromString <$> some (noneOf "{}"))
+
+-- * Parsers (corresponding to EBNF)
+
+core :: (TokenParsing m, Monad m) => m Core
+core = expr
+
+expr :: (TokenParsing m, Monad m) => m Core
+expr = chainl1 atom go where
```
Comment (robrix) [inline semantic-core/src/Data/Core/Parser.hs]: I have an extremely mild preference for writing chains like this infix:

```haskell
expr = atom `chainl1` choice [ … ]
```

I definitely would not hold up merge for this, just wanted to point it out as the idiom isn’t always obvious but might be handy in future.

File: semantic-core/src/Data/Core/Parser.hs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  , ifthenelse
+  , edge
+  , lit
+  , ident
+  , assign
+  , parens expr
+  ]
+
+comp :: (TokenParsing m, Monad m) => m Core
+comp = braces (sconcat <$> sepEndByNonEmpty expr semi)
```
Comment (robrix) [inline semantic-core/src/Data/Core/Parser.hs]: Should this be labelled (`<?>`)?

File: semantic-core/src/Data/Core/Parser.hs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+ifthenelse :: (TokenParsing m, Monad m) => m Core
+ifthenelse = If
+  <$ reserved "if"   <*> core
+  <* reserved "then" <*> core
+  <* reserved "else" <*> core
+  <?> "if-then-else statement"
+
+assign :: (TokenParsing m, Monad m) => m Core
+assign = (:=) <$> try (lvalue <* symbolic '=') <*> core
```
Comment (robrix) [inline semantic-core/src/Data/Core/Parser.hs]: IIRC @ekmett’s advice is that any time you use `try`, you should use `<?>`. I think that applies here.

We could probably also figure out a way to rewrite this without the `try`, but I wouldn’t worry about that unless we started getting really confusing error messages or something.

File: semantic-core/src/Data/Core/Parser.hs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+-- * Parsers (corresponding to EBNF)
+
+core :: (TokenParsing m, Monad m) => m Core
+core = expr
+
+expr :: (TokenParsing m, Monad m) => m Core
+expr = chainl1 atom go where
+  go = choice [ (:.) <$ dot
+              , (:$) <$ notFollowedBy dot
```
Comment (robrix) [inline semantic-core/src/Data/Core/Parser.hs]: argh I still love this so much

File: semantic-core/src/Data/Core/Parser.hs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  [ Bool True  `given` "#true"
+  , Bool False `given` "#false"
+  , Unit       `given` "#unit"
+  , Frame      `given` "#frame"
+  , lambda
+  ] <?> "literal"
+
+lambda :: (TokenParsing m, Monad m) => m Core
+lambda = Lam <$ lambduh <*> name <* arrow <*> core <?> "lambda" where
+  lambduh = symbolic 'λ' <|> symbolic '\\'
```
Comment (robrix) [inline semantic-core/src/Data/Core/Parser.hs]: 😂

File: semantic-core/src/Data/Core/Parser.hs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  arrow   = symbol "→"   <|> symbol "->"
```
Comment (robrix) [inline semantic-core/src/Data/Core/Parser.hs]: I’m still kind of ambivalent about unicode symbols in general, but the pretty-printer is awfully nice and round-tripping seems like a good property, so 👍

File: semantic-core/src/Data/Core/Parser.hs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+-- * Literals
+
+name :: (TokenParsing m, Monad m) => m Name
+name = choice [regular, strpath] <?> "name" where
+  regular = User <$> identifier
+  strpath = Path <$> between (symbolic '"') (symbolic '"') (some $ noneOf "\"")
+
+lit :: (TokenParsing m, Monad m) => m Core
+lit = let x `given` n = x <$ reserved n in choice
```
Comment (robrix) [inline semantic-core/src/Data/Core/Parser.hs]: I always forget you can define infix functions this way, just like any other operator.

File: semantic-core/src/Data/Name.hs [hunk 8]
```diff
@@ -50,6 +54,20 @@ instance Pretty Name where
...
+                , "lexical", "import", "#unit", "load"]
+
+-- | A ‘simple’ character is, loosely defined, a character that is compatible
+-- with identifiers in most ASCII-oriented programming languages. This is defined
+-- as the alphanumeric set plus @$@ and @_@.
+isSimpleCharacter :: Char -> Bool
+isSimpleCharacter = \case
+  '$'  -> True -- common in JS
+  '_'  -> True
+  '?'  -> True -- common in Ruby
```
Comment (robrix) [inline semantic-core/src/Data/Name.hs]: This is probably more reasonable than shoving it into `#{…}` 👍 good call.

File: semantic-core/src/Data/Core/Parser.hs [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+module Data.Core.Parser
+  ( module Text.Trifecta
+  , core
+  , lit
+  , expr
+  , lvalue
+  ) where
+
+-- Consult @doc/grammar.md@ for an EBNF grammar.
```
Comment (robrix) [inline semantic-core/src/Data/Core/Parser.hs]: This file appears to have been omitted.

File: semantic-core/src/Data/Core/Parser.hs [hunk 10]
```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE ExplicitForAll, FlexibleInstances, LambdaCase, MultiParamTypeClasses, OverloadedLists,
+             ScopedTypeVariables #-}
```
Comment (robrix) [inline semantic-core/src/Data/Core/Parser.hs]: I don’t think we’re using any of these extensions any more. No `forall`s, no instances, no `\case`, and the `HashSet` is being defined in `Data.Name` now.

File: semantic-core/src/Data/Core/Parser.hs (hunk 7 above)
Comment (patrickt) [reply semantic-core/src/Data/Core/Parser.hs]: Yup! And you can give them a custom precedence, too!

File: semantic-core/src/Data/Core/Parser.hs (hunk 5 above)
Comment (patrickt) [reply semantic-core/src/Data/Core/Parser.hs]: Based on an Emacs function I have to insert a literal lambda: obviously, `lambda` has a tendency to collide with Lisp operators 😂

File: semantic-core/src/Data/Core/Parser.hs (hunk 3 above)
Comment (patrickt) [reply semantic-core/src/Data/Core/Parser.hs]: That’s good advice. Fixed.

File: semantic-core/src/Data/Core/Parser.hs (hunk 10 above)
Comment (patrickt) [reply semantic-core/src/Data/Core/Parser.hs]: Fixed.

File: docs/core-grammar.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+	      | 'frame'
+	      | lambda
+	      | ident
+
+	lambda ::= ('λ' | '\') ident ('->' | '→') expr
+
+	ident ::= [A-z_] ([A-z0-9_])*
+	        | '#{' [^{}]+ '}'
+	        | '"' [^"]+ '"'
+	```
```
Comment (robrix) [inline docs/core-grammar.md]: I think there’s some weird indentation going on for the actual grammar here.

File: semantic-core/src/Data/Core/Pretty.hs [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+keyword, symbol, strlit, primitive :: AnsiDoc -> AnsiDoc
+keyword = annotate (Pretty.colorDull Pretty.Cyan)
+symbol  = annotate (Pretty.color Pretty.Yellow)
+strlit  = annotate (Pretty.colorDull Pretty.Green)
+primitive = keyword . mappend "#"
+
+type Prec    = Int
+
+data Style = Unicode | Ascii
```
Comment (robrix) [inline semantic-core/src/Data/Core/Pretty.hs]: I dig this 👍

File: semantic-core/src/Data/Core/Pretty.hs [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+  Unicode -> pure $ symbol "λ"
+  Ascii   -> pure $ symbol "\\"
+arrow = ask @Style >>= \case
+  Unicode -> pure $ symbol "→"
+  Ascii   -> pure $ symbol "->"
+
+name :: Name -> AnsiDoc
+name = \case
+  Gen p  -> pretty p
+  Path p -> strlit (Pretty.viaShow p  )
```
Comment (robrix) [inline semantic-core/src/Data/Core/Pretty.hs]:
```suggestion
  Path p -> strlit (Pretty.viaShow p)
```

File: semantic-core/src/Data/Core/Pretty.hs [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+  User n -> encloseIf (needsQuotation n) (symbol "#{") (symbol "}") (pretty n)
```
Comment (robrix) [inline semantic-core/src/Data/Core/Pretty.hs]: ❤️

File: semantic-core/src/Data/Core/Pretty.hs [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+type AnsiDoc = Pretty.Doc Pretty.AnsiStyle
+
+keyword, symbol, strlit, primitive :: AnsiDoc -> AnsiDoc
+keyword = annotate (Pretty.colorDull Pretty.Cyan)
+symbol  = annotate (Pretty.color Pretty.Yellow)
+strlit  = annotate (Pretty.colorDull Pretty.Green)
+primitive = keyword . mappend "#"
+
+type Prec    = Int
```
Comment (robrix) [inline semantic-core/src/Data/Core/Pretty.hs]:
```suggestion
type Prec = Int
```

File: semantic-core/src/Data/Core/Pretty.hs [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+         -> m AnsiDoc
+prettify = \case
+  VarF a -> pure $ name a
+  LetF a -> pure $ keyword "let" <+> name a
+  a :>>$ b -> do
+    prec <- ask @Prec
+    fore <- with 12 a
+    aft  <- with 12 b
+
+    let open  = symbol (if 12 > prec then "{" <> softline else "")
```
Comment (robrix) [inline semantic-core/src/Data/Core/Pretty.hs]: The following three suggestions should be taken or left as a unit:

```suggestion
    let open  = symbol ("{" <> softline)
```

File: semantic-core/src/Data/Core/Pretty.hs [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+        close = symbol (if 12 > prec then softline <> "}" else "")
```
Comment (robrix) [inline semantic-core/src/Data/Core/Pretty.hs]:
```suggestion
        close = symbol (softline <> "}")
```

File: semantic-core/src/Data/Core/Pretty.hs [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+        separator = ";" <> Pretty.line
+        body = fore <> separator <> aft
+
+    pure . Pretty.align $ open <> Pretty.align body <> close
```
Comment (robrix) [inline semantic-core/src/Data/Core/Pretty.hs]:
```suggestion
    pure . Pretty.align $ encloseIf (12 > prec) open close (Pretty.align body)
```

File: semantic-core/src/Data/Core.hs [hunk 19]
```diff
@@ -50,52 +53,16 @@ infixr 1 :>>
...
-data Edge = Lexical | Import
-  deriving (Eq, Ord, Show)
+makeBaseFunctor ''Core
 
-instance Pretty Edge where
-  pretty = pretty . show
+infixl 2 :$$
+infixr 1 :>>$
+infix  3 :=$
+infixl 4 :.$
```
Comment (robrix) [inline semantic-core/src/Data/Core.hs]: Cool!

File: docs/core-grammar.md (hunk 11 above)
Comment (patrickt) [reply docs/core-grammar.md]: Taaaaaaabs.

**Pull request merged**